src/foam/nanos/auth/CapabilityAuthService.js

```diff
@@ -163,81 +163,36 @@ foam.CLASS({
-              if ( result ) {
-                ((Map<String, Boolean>) getCache()).put(userKey, result);
-                return result;
-              }
-            } else if ( cap != null && agent != null && agent.getId() != user.getId() ) {
-              // if the agent is in the context and is not the same as user, check 
-              // if the agent has capability
-              userPredicate = AND(
-                INSTANCE_OF(AgentCapabilityJunction.class),
-                EQ(UserCapabilityJunction.SOURCE_ID, agent.getId()),
-                EQ(AgentCapabilityJunction.EFFECTIVE_USER, user.getId())
-              );
-              if ( userCapabilityJunctionDAO.find(
-                AND(
-                  userPredicate,
-                  capabilityScope,
-                  EQ(UserCapabilityJunction.TARGET_ID, cap.getId()),
-                  EQ(UserCapabilityJunction.STATUS, CapabilityJunctionStatus.GRANTED)
-                )) != null )
-                result = true;
-                // if the user has the permission, store this in the cache and return the result
-                // otherwise, move on to the 2nd part of the check
-                if ( result ) {
-                  if ( agentKey != null ) ((Map<String, Boolean>) getCache()).put(agentKey, result);
-                  return result;
-                }
-            }
-          }
-
-          // 2. check if the user has a capability that grants the permission
-          userPredicate = EQ(UserCapabilityJunction.SOURCE_ID, user.getId());
           AbstractPredicate predicate = new AbstractPredicate(x) {
             @Override
             public boolean f(Object obj) {
               UserCapabilityJunction ucj = (UserCapabilityJunction) obj;
               Capability c = (Capability) capabilityDAO.find(ucj.getTargetId());
-              if ( ucj.getStatus() == CapabilityJunctionStatus.GRANTED && 
+              if ( ucj.getStatus() == CapabilityJunctionStatus.GRANTED &&
                    c != null && ! c.isDeprecated(x) && c.implies(x, permission) ) {
                 return true;
               }
               return false;
             }
           };
+
+          // 1. check if the user has a capability that grants the permission
+          Predicate userPredicate = EQ(UserCapabilityJunction.SOURCE_ID, user.getId());
           if ( userCapabilityJunctionDAO.find(AND(userPredicate, capabilityScope, predicate)) != null ) {
             result = true;
           }
           // Add the result to the cache
+          (( Map<String, Boolean> ) getCache()).put(userKey, result);
           if ( result ) {
-            (( Map<String, Boolean> ) getCache()).put(userKey, result);
             return true;
           }
 
-          // 3. check if the agent has a capability that grants the permission
-          if ( agent != null ) {
+          // 2. check if the agent has a capability that grants the permission
```

Comment:
suggest a comment update:
//# Check if a ucj implies the subject.realUser(agent) has this permission